### PR TITLE
Add client services for reading state and submitting transactions 

### LINKED ISSUE #22: Client: Create module to make API requests
Getting state information from the Sawtooth REST API will be non-trivial, and should be encapsulated in one or more modules.

First, it should be able to make a `/state` request specific to one type of resource, or one particular resource. Pages using this module should have no knowledge of how addressing works. They would only have to specify resource type and (if applicable) identifier.

Second, it must be able to decode the bytes information that comes back into JavaScript objects or arrays. Pages using this module should have no knowledge of the encoding scheme used for on-chain data.

Finally, it must be able to encode and submit transactions, including waiting on the results to be committed, and fetching those results.

The rest of the client app should be able to use this module with the same level of abstraction as they would a typical REST API, but with function calls instead of HTTP requests.

### PR DESCRIPTION
**Pull request by delventhalz**
Closes #22. Adds two modules to assist the client in interacting with the Sawtooth blockchain:

## services/requests

Includes four methods for fetching state. Each operates in a vaguely RESTful manner, retrieving all resources if no identifier is passed in, or one specific resource if there is an id. All are Promise based.

#### getCollections

- **getCollections()** - fetch all collections
- **getCollections(publicKey)** - fetch one collection

#### getMoji

- **getMoji()** - fetch all moji
- **getMoji(address)** - fetch one moji
- **getMoji({ owner: publicKey })** - fetch all moji owned by one collection
- **getMoji({ owner: publicKey, dna: dnaString })** - fetch one moji

#### getOffers

- **getOffers()** - fetch all offers
- **getOffers(address)** - fetch one offers
- **getOffers({ owner: publicKey })** - fetch all offers owned by one collection
- **getOffers({ owner: publicKey, moji: [addressesOrDna] })** - fetch one offer

#### getSires

- **getSires()** - fetch all sires (returns whole moji, not just address)
- **getSires(publicKey)** - fetch the sire belonging to a particular owner

## services/transactions

Has three methods which can be used to build and submit transactions.

#### createKeys()

Creates a new private/public key pair, and returns them in an object with the keys `privateKey` and `publicKey`. Both are represented as hexadecimal strings.

#### getPublicKey(privateKey)

Takes a private key as a hex string, and returns its public key, also as a hex string.

#### getSubmitter(privateKey)

Takes a private key and returns a function which can be used to submit transactions. This function accepts two parameters. First, the payload or payloads you are going to submit (an object or an array of objects). The second is an optional boolean, specifying whether the returned Promise should wait to resolve until the transaction(s) are committed. It defaults to true.

### Example

```javascript
const { privateKey, publicKey } = transactions.createKeys();

const submit = transactions.getSubmitter(privateKey);

submit({ action: 'CREATE_COLLECTION' })
  .then(() => requests.getCollections(publicKey)
  .then(myCollection => console.log(myCollection)
```

## Testing

I am not entirely sure how ES6 modules get plugged into the front end yet, so there is probably an easier way to test it, but I tested these files with Node. This required making some modifications:

- `import XXXX from YYYY;` -> `const XXXX = require(YYYY);`
- `api/XXXX` -> `http://localhost:8000/api/XXXX`
- `export` -> delete 
- add all exported functions to Node exports: `module.exports = { XXXX, YYYY };`

Once you make those swaps, you can:

```bash
cd code/part-two/
docker-compose up --build
```

Then in a new tab:

```bash
cd client/
npm install
node
```

Then from the node console:

```javascript
const requests = require('./source/services/requests');
const transactions = require('./source/services/transactions');
```

Then you can mess around with the two services to your heart's content. Keep in mind you won't be able to successfully send any transactions other than create collection, or fetch any data other than collections because the TP doesn't have those features yet. Once PR #37 is merged, you will be able to fetch Moji data as well.

### CODE REVIEW

File: code/part-two/client/source/services/transactions.js [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    url: 'api/batches',
+    data: batchList,
+    headers: { 'Content-Type': 'application/octet-stream' }
+  })
+    .then(({ data }) => {
+      if (!shouldWait) {
+        return true;
+      }
+
+      return axios.get(data.link + '&wait')
```
Comment (therobinkim) [inline code/part-two/client/source/services/transactions.js]: minor suggestion: end the `.then(...)` block from ln 98 right after ln 103, and then attach the `.then(...) on ln 104 afterwards

File: code/part-two/client/source/services/transactions.js [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+// Encode a Batch in a BatchList
+const encodeBatch = batch => {
+  return BatchList.encode({ batches: [batch] }).finish();
+};
+
+// Submits an encoded BatchList to the API
+const submitBatchList = (batchList, shouldWait) => {
+  return axios({
+    method: 'POST',
+    url: 'api/batches',
```
Comment (therobinkim) [inline code/part-two/client/source/services/transactions.js]: do we need a `/` at the beginning of the url?

Comment (delventhalz) [reply code/part-two/client/source/services/transactions.js]: I'm unsure actually. When I tested it I was subbing in the full URLs. In other front-end code I've written though (non-axios), I've started relative URLs without a leading slash:

https://github.com/hyperledger/sawtooth-supply-chain/blob/master/asset_client/src/services/api.js#L23

Comment (therobinkim) [reply code/part-two/client/source/services/transactions.js]: we can leave it for now and I can let you know if I end up running into any issues ¯\_(ツ)_/¯

**Pull request merged**